src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -2676,23 +2676,34 @@ private void ReportCantConvertLambdaReturn(SyntaxNode syntax, DiagnosticBag diag
             var lambda = this.ContainingMemberOrLambda as LambdaSymbol;
             if ((object)lambda != null)
             {
+                Location location = GetLocationForDiagnostics(syntax);
                 if (IsInAsyncMethod())
                 {
                     // Cannot convert async {0} to intended delegate type. An async {0} may return void, Task or Task<T>, none of which are convertible to '{1}'.
                     Error(diagnostics, ErrorCode.ERR_CantConvAsyncAnonFuncReturns,
-                        syntax,
+                        location,
                         lambda.MessageID.Localize(), lambda.ReturnType);
                 }
                 else
                 {
                     // Cannot convert {0} to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type
                     Error(diagnostics, ErrorCode.ERR_CantConvAnonMethReturns,
-                        syntax,
+                        location,
                         lambda.MessageID.Localize());
                 }
             }
         }
 
+        private static Location GetLocationForDiagnostics(SyntaxNode node)
+        {
+            if (node is LambdaExpressionSyntax lambda)
+            {
+                return Location.Create(lambda.SyntaxTree, Text.TextSpan.FromBounds(lambda.SpanStart, lambda.ArrowToken.Span.End));
```

Comment:
>return Location.Create(lambda.SyntaxTree, Text.TextSpan.FromBounds(lambda.SpanStart, lambda.ArrowToken.Span.End)); [](start = 16, length = 114)

We already have LambdaSymbol.DiagnosticLocation property which has different implementation and handles delegate ... syntax. Perhaps we should unify/share/reuse the implementation.

Reply:
Yes, but it only squiggles the arrow. That's appropriate when there is an error in the lambda, but doesn't seem great for this scenario (consuming the lambda). If it's ok, I'll leave as a separate implementation.